File: backend/services/notification_service.py
from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


class NotificationService:
    """Zentraler Notification-Service für Messenger- und System-Benachrichtigungen.

    Implementiert strikte Outgoing Echo Prevention und Empfänger-Filterung:
    - Wenn die KI eine Nachricht im Namen des angemeldeten Benutzers versendet,
      darf für den Sender-Account KEINE Push-Benachrichtigung oder Unread-Badge getriggert werden.
    - Push/Alerts und Unread-Badges gibt es ausschließlich für den Empfänger bei
      ungesehenen Nachrichten (recipient_id == current_user && !is_read).
    """

    @staticmethod
    def is_outgoing_echo(
        *,
        sender_user_id: int | str | None,
        current_user_id: int | str | None,
    ) -> bool:
        """Prüft strikt, ob die Aktion vom aktuellen Benutzer selbst (oder in seinem Namen via KI/Worker)
        ausgelöst wurde (Outgoing Echo).
        """
        if sender_user_id is None or current_user_id is None:
            return False
        try:
            return int(sender_user_id) == int(current_user_id)
        except (ValueError, TypeError):
            pass

        s_str = str(sender_user_id).strip()
        c_str = str(current_user_id).strip()
        if not s_str or not c_str:
            return False
        if s_str == c_str:
            return True

        # Unterstützt Worker/AI/Actor-Präfixe wie "worker:42", "ai:42", "actor:42", "user:42"
        for prefix in ("worker:", "ai:", "actor:", "user:"):
            if s_str.startswith(prefix):
                sub = s_str[len(prefix) :].strip()
                try:
                    if int(sub) == int(current_user_id):
                        return True
                except (ValueError, TypeError):
                    if sub == c_str:
                        return True

        return False

    @staticmethod
    def should_notify(
        *,
        recipient_id: int | str | None,
        current_user_id: int | str | None,
        is_read: bool = False,
        sender_user_id: int | str | None = None,
        is_group: bool = False,
        is_control: bool = False,
        control_type: str | None = None,
        has_active_foreground_client: bool = False,
    ) -> bool:
        """Prüft, ob für den aktuellen Benutzer ein Push-Alert oder Unread-Badge ausgelöst werden darf.

        Strikte Invarianten:
        - `sender_user_id == current_user_id`: Absender wird niemals benachrichtigt (Outgoing Echo Prevention).
        - `is_control`: Interne Steuernachrichten (Read Receipts, Delivery Receipts, Typing) triggern keine Alerts/Badges.
        - `is_read`: Gesehene oder bereits gelesene Nachrichten triggern keinen Alert (z. B. Chat-Fokus).
        - `has_active_foreground_client`: Bei aktiver Foreground-WebSocket/SSE-Verbindung keine doppelten OS-Pushes.
        - `recipient_id == current_user_id`: Push/Alerts nur wenn der Benutzer auch wirklich der Empfänger ist.
        """
        if current_user_id is None:
            return False

        # 1. Outgoing Echo Prevention: Sender darf niemals über eigene Aktionen benachrichtigt werden
        # (gilt ausnahmslos auch für Aktionen via KI/Worker im Namen des Nutzers)
        if NotificationService.is_outgoing_echo(
            sender_user_id=sender_user_id, current_user_id=current_user_id
        ):
            return False

        try:
            cur_id = int(current_user_id)
        except (ValueError, TypeError):
            cur_id = None

        # 2. Interne Steuernachrichten (z. B. read_receipt, delivery_receipt, edit_message, delete_message)
        if is_control or (control_type and control_type not in ("message", "normal")):
            return False

        # 3. Bereits gesehene oder gelesene Nachrichten (z. B. Chatfenster aktiv im Fokus) -> kein Alert
        if is_read:
            return False

        # 4. Aktive Verbindung im Vordergrund -> OS-Pushes unterdrücken, um Doppelung zu verhindern
        if has_active_foreground_client:
            return False

        # 5. Empfängerprüfung für Direktnachrichten
        if recipient_id is not None:
            if cur_id is not None:
                try:
                    return int(recipient_id) == cur_id
                except (ValueError, TypeError):
                    return False
            return str(recipient_id).strip() == str(current_user_id).strip()

        # 6. Gruppen-Nachrichten (wenn recipient_id nicht definiert ist)
        if is_group:
            return True

        # Falls recipient_id nicht definiert ist (Legacy / Broadcast):
        # Nur benachrichtigen, wenn der Sender nicht der aktuelle Nutzer ist
        if sender_user_id is not None:
            return not NotificationService.is_outgoing_echo(
                sender_user_id=sender_user_id, current_user_id=current_user_id
            )

        return False

    @staticmethod
    def should_notify_group(
        *,
        group_member_ids: list[int | str] | set[int | str],
        current_user_id: int | str | None,
        is_read: bool = False,
        sender_user_id: int | str | None = None,
        is_control: bool = False,
        control_type: str | None = None,
        has_active_foreground_client: bool = False,
    ) -> bool:
        """Prüft Gruppen-Nachrichten: Outgoing Echo Prevention, Steuernachrichten und Gruppenmitgliedschaft."""
        if current_user_id is None:
            return False

        # 1. Outgoing Echo Prevention
        if NotificationService.is_outgoing_echo(
            sender_user_id=sender_user_id, current_user_id=current_user_id
        ):
            return False

        # 2. Steuernachrichten ausschließen
        if is_control or (control_type and control_type not in ("message", "normal")):
            return False

        # 3. Bereits gelesen / Fokus
        if is_read:
            return False

        # 4. Aktiver Foreground-Client
        if has_active_foreground_client:
            return False

        try:
            cur_id = int(current_user_id)
        except (ValueError, TypeError):
            cur_id = None

        cur_str = str(current_user_id).strip()
        for mid in group_member_ids:
            if cur_id is not None:
                try:
                    if int(mid) == cur_id:
                        return True
                except (ValueError, TypeError):
                    pass
            if str(mid).strip() == cur_str:
                return True

        return False

    @staticmethod
    def sanitize_push_payload(
        payload: dict[str, Any],
        *,
        is_e2ee: bool = True,
        privacy_mode: bool = True,
    ) -> dict[str, Any]:
        """Bereinigt Push-Payloads (WebPush / FCM) gegen Privacy-Leaks.

        Strikte Zero-Knowledge- und Datenschutz-Garantien:
        - NIEMALS Klartext-Nachrichteninhalte (body, text, message_text, preview) in Push-Payloads leaken,
          wenn E2EE oder Privatsphäre-Filter aktiv sind.
        - Sensible kryptographische Details (ciphertext_envelope, secret, key, token) werden entfernt.
        - Bei E2EE wird ein generischer, sicherer Text verwendet ('Neue verschlüsselte Nachricht').
        """
        sanitized = dict(payload)

        # Gefährliche Klartext- und Sensitiv-Felder
        leak_keys = [
            "text",
            "message_text",
            "message",
            "msg",
            "plaintext",
            "content",
            "body",
            "preview",
            "ciphertext_envelope",
            "raw_payload",
            "secret",
            "token",
            "key",
            "password",
            "attachments",
            "attachment",
            "private_key",
            "secret_key",
            "channel_key",
            "iv",
            "auth_tag",
            "salt",
        ]

        if is_e2ee or privacy_mode:
            # Jeglichen Klartext entfernen
            for k in leak_keys:
                sanitized.pop(k, None)

            # Generischen, sicheren Text garantieren
            is_group = bool(payload.get("is_group", False))
            chat_name = payload.get("chat_name") or payload.get("sender_name")
            if chat_name:
                sanitized["title"] = f"Neue Nachricht: {chat_name}" if not is_group else f"Neue Nachricht in „{chat_name}“"
            else:
                sanitized["title"] = "Neue Nachricht"

            sanitized["body"] = (
                "Neue Nachricht in der Gruppe."
                if is_group
                else "Du hast eine neue verschlüsselte Nachricht erhalten."
            )
            sanitized["is_e2ee"] = True
            sanitized["privacy_filtered"] = True
        else:
            # Auch im normalen Modus niemals kryptographische Secrets leaken
            for k in (
                "ciphertext_envelope",
                "secret",
                "token",
                "key",
                "password",
                "private_key",
                "secret_key",
                "channel_key",
                "iv",
                "auth_tag",
                "salt",
            ):
                sanitized.pop(k, None)

        return sanitized

    @staticmethod
    def prepare_push_dispatch(
        *,
        target_user_id: int | str,
        sender_user_id: int | str | None = None,
        title: str,
        body: str | None = None,
        is_e2ee: bool = True,
        privacy_mode: bool = True,
        is_control: bool = False,
        control_type: str | None = None,
        has_active_foreground_connection: bool = False,
        extra_data: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        """Erzeugt einen validierten, bereinigten WebPush/FCM Push-Dispatch-Auftrag.

        Liefert None zurück, wenn:
        - target_user_id == sender_user_id (Outgoing Echo Prevention)
        - is_control True ist (keine Pushes für Lesequittungen / interne Steuersignale)
        - has_active_foreground_connection True ist (keine doppelten OS-Pushes bei aktivem Client)
        """
        # 1. Outgoing Echo Prevention: Sender darf niemals benachrichtigt werden
        if NotificationService.is_outgoing_echo(
            sender_user_id=sender_user_id, current_user_id=target_user_id
        ):
            return None

        # 2. Steuersignale triggern keine OS-Pushes
        if is_control or (control_type and control_type not in ("message", "normal")):
            return None

        # 3. Vordergrund-Prüfung: Aktive Verbindung unterdrückt OS-Push zur Vermeidung von Doppelung
        if has_active_foreground_connection:
            return None

        raw_payload = {
            "title": title,
            "body": body or "Neue Benachrichtigung",
            "target_user_id": target_user_id,
            "sender_user_id": sender_user_id,
            **(extra_data or {}),
        }

        return NotificationService.sanitize_push_payload(
            raw_payload, is_e2ee=is_e2ee, privacy_mode=privacy_mode
        )

    @staticmethod
    def dispatch_push(
        *,
        target_user_id: int | str,
        sender_user_id: int | str | None = None,
        title: str,
        body: str | None = None,
        is_e2ee: bool = True,
        privacy_mode: bool = True,
        is_control: bool = False,
        control_type: str | None = None,
        has_active_foreground_connection: bool = False,
        extra_data: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        """Zentraler Push-Dispatcher: Filtert, bereinigt und plant WebPush/FCM-Payloads."""
        payload = NotificationService.prepare_push_dispatch(
            target_user_id=target_user_id,
            sender_user_id=sender_user_id,
            title=title,
            body=body,
            is_e2ee=is_e2ee,
            privacy_mode=privacy_mode,
            is_control=is_control,
            control_type=control_type,
            has_active_foreground_connection=has_active_foreground_connection,
            extra_data=extra_data,
        )
        if payload is None:
            return None
        logger.debug("Push-Dispatch vorbereitet für User %s: %s", target_user_id, payload.get("title"))
        return payload

File: frontend/src/services/notificationService.ts
/**
 * NotificationService
 *
 * Verwaltet die Filterlogik für Geräte-, Push- und Unread-Benachrichtigungen.
 *
 * Strikte Sicherheits- und UX-Invarianten:
 * 1. Outgoing Echo Prevention: Wenn der Benutzer (oder ein KI-Worker in seinem Namen)
 *    eine Nachricht versendet, darf für den Sender-Account KEINE Push-Benachrichtigung
 *    oder Unread-Badge getriggert werden.
 * 2. Empfänger-Filterung: Push/Alerts gibt es ausschließlich für den tatsächlichen
 *    Empfänger bei ungesehenen Nachrichten (`recipient_id == current_user && !is_read`).
 */

export interface NotificationFilterParams {
  recipientId?: number | string | null
  currentUserId?: number | string | null
  isRead?: boolean
  senderUserId?: number | string | null
  isGroup?: boolean
  isControl?: boolean
  controlType?: string | null
  hasActiveForegroundConnection?: boolean
}

export interface PushPayloadSanitizeParams {
  title?: string
  body?: string
  isGroup?: boolean
  chatName?: string
  senderName?: string
  isE2ee?: boolean
  privacyMode?: boolean
  extraData?: Record<string, unknown>
}

export interface PreparePushDispatchParams {
  targetUserId: number | string
  senderUserId?: number | string | null
  title: string
  body?: string
  isE2ee?: boolean
  privacyMode?: boolean
  isControl?: boolean
  controlType?: string | null
  hasActiveForegroundConnection?: boolean
  extraData?: Record<string, unknown>
}

export class NotificationService {
  /**
   * Prüft strikt, ob die Aktion vom aktuellen Benutzer selbst (oder in seinem Namen via KI/Worker)
   * ausgelöst wurde (Outgoing Echo).
   */
  static isOutgoingEcho(
    senderUserId: number | string | null | undefined,
    currentUserId: number | string | null | undefined
  ): boolean {
    if (senderUserId == null || currentUserId == null) return false
    const sndStr = String(senderUserId).trim()
    const curStr = String(currentUserId).trim()
    if (!sndStr || !curStr) return false
    if (sndStr === curStr) return true

    const sndNum = Number(sndStr)
    const curNum = Number(curStr)
    if (!isNaN(sndNum) && !isNaN(curNum) && sndNum === curNum) {
      return true
    }

    for (const prefix of ['worker:', 'ai:', 'actor:', 'user:']) {
      if (sndStr.startsWith(prefix)) {
        const sub = sndStr.slice(prefix.length).trim()
        const subNum = Number(sub)
        if (!isNaN(subNum) && !isNaN(curNum) && subNum === curNum) {
          return true
        }
        if (sub === curStr) {
          return true
        }
      }
    }

    return false
  }

  /**
   * Prüft strikt, ob eine Benachrichtigung / Unread-Badge getriggert werden darf.
   */
  static shouldNotify({
    recipientId,
    currentUserId,
    isRead = false,
    senderUserId,
    isGroup = false,
    isControl = false,
    controlType,
    hasActiveForegroundConnection = false,
  }: NotificationFilterParams): boolean {
    if (currentUserId == null) {
      return false
    }

    // 1. Outgoing Echo Prevention: Sender darf niemals benachrichtigt werden
    if (NotificationService.isOutgoingEcho(senderUserId, currentUserId)) {
      return false
    }

    // 2. Interne Steuersignale (read receipts, delivery receipts, typing, edit, delete) ausschließen
    if (isControl || (controlType && controlType !== 'message' && controlType !== 'normal')) {
      return false
    }

    // 3. Bereits gesehene/gelesene Nachricht (z. B. Chatfenster aktiv geöffnet) -> kein Alert/Badge
    if (isRead) {
      return false
    }

    // 4. Bei aktiver Verbindung im Vordergrund keine doppelten OS-Pushes auslösen
    if (hasActiveForegroundConnection) {
      return false
    }

    // 5. Strikte Empfängerprüfung für Direktnachrichten
    if (recipientId != null) {
      const recNum = Number(recipientId)
      const curNum = Number(currentUserId)
      if (!isNaN(recNum) && !isNaN(curNum)) {
        return recNum === curNum
      }
      return String(recipientId).trim() === String(currentUserId).trim()
    }

    // 6. Gruppen-Nachrichten (wenn recipientId nicht gesetzt ist)
    if (isGroup) {
      return true
    }

    // Falls recipientId nicht im Event enthalten ist (z. B. Legacy Broadcast):
    // Alert nur auslösen, wenn senderUserId nicht der aktuelle Benutzer ist
    if (senderUserId != null) {
      return !NotificationService.isOutgoingEcho(senderUserId, currentUserId)
    }

    return false
  }

  /**
   * Prüft Gruppen-Nachrichten auf Mitgliedschaft, Outgoing Echo Prevention und Steuernachrichten.
   */
  static shouldNotifyGroup({
    groupMemberIds,
    currentUserId,
    isRead = false,
    senderUserId,
    isControl = false,
    controlType,
    hasActiveForegroundConnection = false,
  }: {
    groupMemberIds: (number | string)[]
    currentUserId?: number | string | null
    isRead?: boolean
    senderUserId?: number | string | null
    isControl?: boolean
    controlType?: string | null
    hasActiveForegroundConnection?: boolean
  }): boolean {
    if (currentUserId == null) return false
    if (NotificationService.isOutgoingEcho(senderUserId, currentUserId)) return false
    if (isControl || (controlType && controlType !== 'message' && controlType !== 'normal')) return false
    if (isRead) return false
    if (hasActiveForegroundConnection) return false

    const curNum = Number(currentUserId)
    const curStr = String(currentUserId).trim()

    return groupMemberIds.some((id) => {
      const num = Number(id)
      if (!isNaN(num) && !isNaN(curNum) && num === curNum) return true
      return String(id).trim() === curStr
    })
  }

  /**
   * Bereinigt Push-Payloads gegen Privacy-Leaks (Klartext, Metadaten, Secrets).
   */
  static sanitizePushPayload({
    title,
    body,
    isGroup = false,
    chatName,
    senderName,
    isE2ee = true,
    privacyMode = true,
    extraData,
  }: PushPayloadSanitizeParams): Record<string, unknown> {
    const sanitized: Record<string, unknown> = { ...(extraData || {}) }

    // Bereinige gefährliche Klartextfelder
    const leakKeys = [
      'text',
      'message_text',
      'message',
      'msg',
      'plaintext',
      'content',
      'body',
      'preview',
      'ciphertext_envelope',
      'raw_payload',
      'secret',
      'token',
      'key',
      'password',
      'attachments',
      'attachment',
      'private_key',
      'secret_key',
      'channel_key',
      'iv',
      'auth_tag',
      'salt',
    ]
    for (const k of leakKeys) {
      delete sanitized[k]
    }

    if (isE2ee || privacyMode) {
      const name = chatName || senderName
      sanitized.title = name ? (isGroup ? `Neue Nachricht in „${name}“` : `Neue Nachricht: ${name}`) : 'Neue Nachricht'
      sanitized.body = isGroup ? 'Neue Nachricht in der Gruppe.' : 'Du hast eine neue verschlüsselte Nachricht erhalten.'
      sanitized.is_e2ee = true
      sanitized.privacy_filtered = true
    } else {
      sanitized.title = title || 'Neue Benachrichtigung'
      sanitized.body = body || 'Neue Nachricht erhalten.'
    }

    return sanitized
  }

  /**
   * Erzeugt einen validierten, bereinigten Push-Dispatch-Auftrag oder null, wenn unterdrückt.
   */
  static preparePushDispatch({
    targetUserId,
    senderUserId,
    title,
    body,
    isE2ee = true,
    privacyMode = true,
    isControl = false,
    controlType,
    hasActiveForegroundConnection = false,
    extraData,
  }: PreparePushDispatchParams): Record<string, unknown> | null {
    if (NotificationService.isOutgoingEcho(senderUserId, targetUserId)) {
      return null
    }
    if (isControl || (controlType && controlType !== 'message' && controlType !== 'normal')) {
      return null
    }
    if (hasActiveForegroundConnection) {
      return null
    }

    return NotificationService.sanitizePushPayload({
      title,
      body,
      isE2ee,
      privacyMode,
      extraData: {
        target_user_id: targetUserId,
        sender_user_id: senderUserId,
        ...extraData,
      },
    })
  }

  /**
   * Zentraler Push-Dispatcher: Filtert, bereinigt und plant Push-Payloads.
   */
  static dispatchPush(params: PreparePushDispatchParams): Record<string, unknown> | null {
    return NotificationService.preparePushDispatch(params)
  }
}


